### PR TITLE
Pretty Printing: Trace

### LINKED ISSUE #9044: pretty printing: `Trace` could be improved (and LaTeX)
````
In [69]: X = ImmutableMatrix([[1, 2], [3, 4]])

In [70]: pprint(X)
⎡1  2⎤
⎢    ⎥
⎣3  4⎦

In [71]: w = Trace(X) + Trace(2*X)

In [72]: pprint(w)
Trace(Matrix([ + Trace(Matrix([
   [1, 2],          [2, 4],    
  [3, 4]]))        [6, 8]]))   
````

I suppose `pprint(Trace(X))` should be:
````
  ⎛⎡1  2⎤⎞ 
tr⎜⎢    ⎥⎟ 
  ⎝⎣3  4⎦⎠ 
````

### PR DESCRIPTION
**Pull request by tokencolour**
Attempts to fix #9044

### CODE REVIEW

**Comment by cbm755** [comment]
I will try to review later.  But for sure it will need test(s).

**Comment by tokencolour** [comment]
I have added the tests. @cbm755 , Should you find time, please review it. Thank you.

File: sympy/printing/pretty/tests/test_pretty.py [hunk 1]
```diff
@@ -2627,6 +2627,39 @@ def test_Adjoint():
     assert upretty(Transpose(Adjoint(X))) == \
         u("    T\n⎛ †⎞ \n⎝X ⎠ ")
 
+def test_pretty_Trace_issue_9044():
+    X = ImmutableMatrix([[1, 2], [3, 4]])
+    Y = ImmutableMatrix([[2, 4], [6, 8]])
```
Comment (cbm755) [inline sympy/printing/pretty/tests/test_pretty.py]: Just `Matrix` here

File: sympy/printing/pretty/tests/test_pretty.py [hunk 2]
```diff
@@ -1,10 +1,10 @@
 # -*- coding: utf-8 -*-
 from sympy import (
     Add, And, Basic, Derivative, Dict, Eq, Equivalent, FF,
-    FiniteSet, Function, Ge, Gt, I, Implies, Integral,
+    FiniteSet, Function, Ge, Gt, I, Implies, ImmutableMatrix, Integral,
```
Comment (cbm755) [inline sympy/printing/pretty/tests/test_pretty.py]: Not needed

**Comment by cbm755** [comment]
Test looks good.  Few minor comments above.

This is getting there, but vertical alignment is still not right:
````
>>> Z = Matrix(7,7, lambda i,j: i+j)
>>> pprint(Trace(Z))
  ⎛⎡0  1  2  3  4   5   6 ⎤⎞
  ⎜⎢                      ⎥⎟
  ⎜⎢1  2  3  4  5   6   7 ⎥⎟
  ⎜⎢                      ⎥⎟
  ⎜⎢2  3  4  5  6   7   8 ⎥⎟
  ⎜⎢                      ⎥⎟
  ⎜⎢3  4  5  6  7   8   9 ⎥⎟
  ⎜⎢                      ⎥⎟
  ⎜⎢4  5  6  7  8   9   10⎥⎟
tr⎜⎢                      ⎥⎟
  ⎜⎢5  6  7  8  9   10  11⎥⎟
  ⎜⎢                      ⎥⎟
  ⎝⎣6  7  8  9  10  11  12⎦⎠
````

**Comment by tokencolour** [comment]
This makes it right. @cbm755, Please take a look at it. Should I add more tests?

**Comment by cbm755** [comment]
@tokencolour maybe I haven't thought carefully enough about what this *should* do: I've been assuming `tr` should be printed in the middle vertically: but I'm not sure that assumption is correct.

Can I suggest studying #9316?

I think the question comes down to "what is the baseline" for a matrix?

Here's some examples I was just playing with, that helped me understand `sin`s pretty behaviour:
````
In [63]: pprint(sin(F) + Trace(Z))
                       ⎛⎡0  1  2  3  4⎤⎞
                       ⎜⎢             ⎥⎟
                       ⎜⎢1  2  3  4  5⎥⎟
   ⎛        1    ⎞     ⎜⎢             ⎥⎟
sin⎜x + ─────────⎟ + tr⎜⎢2  3  4  5  6⎥⎟
   ⎜          1  ⎟     ⎜⎢             ⎥⎟
   ⎜    y + ─────⎟     ⎜⎢3  4  5  6  7⎥⎟
   ⎜            1⎟     ⎜⎢             ⎥⎟
   ⎜        z + ─⎟     ⎝⎣4  5  6  7  8⎦⎠
   ⎝            w⎠                      

In [64]: pprint(2**Trace(Z))
   ⎛⎡0  1  2  3  4⎤⎞
   ⎜⎢             ⎥⎟
   ⎜⎢1  2  3  4  5⎥⎟
   ⎜⎢             ⎥⎟
 tr⎜⎢2  3  4  5  6⎥⎟
   ⎜⎢             ⎥⎟
   ⎜⎢3  4  5  6  7⎥⎟
   ⎜⎢             ⎥⎟
   ⎝⎣4  5  6  7  8⎦⎠
2                   

In [65]: pprint(2**sin(F))
    ⎛        1    ⎞
 sin⎜x + ─────────⎟
    ⎜          1  ⎟
    ⎜    y + ─────⎟
    ⎜            1⎟
    ⎜        z + ─⎟
    ⎝            w⎠
2                  

In [66]: pprint(2**sin(G))
    ⎛                   ⎛x⎞⎞
    ⎜                sin⎜─⎟⎟
    ⎜        1          ⎝2⎠⎟
 sin⎜x + ───────── + ──────⎟
    ⎜          1       t   ⎟
    ⎜    y + ─────         ⎟
    ⎜            1         ⎟
    ⎜        z + ─         ⎟
    ⎝            w         ⎠
2                           
````

So now consider unevaluated determinant (which is at least somewhat similar to our Trace):
````
In [87]: pprint(2 + det(MatPow(Z,3)))
    │                 3│
2 + │⎛⎡0  1  2  3  4⎤⎞ │
    │⎜⎢             ⎥⎟ │
    │⎜⎢1  2  3  4  5⎥⎟ │
    │⎜⎢             ⎥⎟ │
    │⎜⎢2  3  4  5  6⎥⎟ │
    │⎜⎢             ⎥⎟ │
    │⎜⎢3  4  5  6  7⎥⎟ │
    │⎜⎢             ⎥⎟ │
    │⎝⎣4  5  6  7  8⎦⎠ │

In [88]: pprint(sin(det(MatPow(Z,3))))
   ⎛│                 3│⎞
sin⎜│⎛⎡0  1  2  3  4⎤⎞ │⎟
   ⎜│⎜⎢             ⎥⎟ │⎟
   ⎜│⎜⎢1  2  3  4  5⎥⎟ │⎟
   ⎜│⎜⎢             ⎥⎟ │⎟
   ⎜│⎜⎢2  3  4  5  6⎥⎟ │⎟
   ⎜│⎜⎢             ⎥⎟ │⎟
   ⎜│⎜⎢3  4  5  6  7⎥⎟ │⎟
   ⎜│⎜⎢             ⎥⎟ │⎟
   ⎝│⎝⎣4  5  6  7  8⎦⎠ │⎠
````

Also:
````
In [91]: pprint(2 + det(MatMul(Z,2)))
2 + 32⋅│⎡0  1  2  3  4⎤│
       │⎢             ⎥│
       │⎢1  2  3  4  5⎥│
       │⎢             ⎥│
       │⎢2  3  4  5  6⎥│
       │⎢             ⎥│
       │⎢3  4  5  6  7⎥│
       │⎢             ⎥│
       │⎣4  5  6  7  8⎦│
````

Does that mean we should have `tr` at the top?  (its not rhetorical, its a genuine question).

**Comment by cbm755** [comment]
CC a few people who (I think) are interested in pretty printing: @hargup @aktech  @asmeurer  

Basic question about baseline of a matrix, specifically:
````
>>> pprint(Trace(Z))    # A
  ⎛⎡0  1  2  3  4⎤⎞
  ⎜⎢             ⎥⎟
  ⎜⎢1  2  3  4  5⎥⎟
  ⎜⎢             ⎥⎟
tr⎜⎢2  3  4  5  6⎥⎟
  ⎜⎢             ⎥⎟
  ⎜⎢3  4  5  6  7⎥⎟
  ⎜⎢             ⎥⎟
  ⎝⎣4  5  6  7  8⎦⎠
````
or
````
>>> pprint(Trace(Z))     # B
tr⎛⎡0  1  2  3  4⎤⎞
  ⎜⎢             ⎥⎟
  ⎜⎢1  2  3  4  5⎥⎟
  ⎜⎢             ⎥⎟
  ⎜⎢2  3  4  5  6⎥⎟
  ⎜⎢             ⎥⎟
  ⎜⎢3  4  5  6  7⎥⎟
  ⎜⎢             ⎥⎟
  ⎝⎣4  5  6  7  8⎦⎠
````
(obviously A looks better here, but other things don't work that way, see discussion above)

**Comment by tokencolour** [comment]
@cbm755 , I read the issue #9316 and I understand the problem of the position of the baseline. Furthermore, the tests have failed. I can't figure out why.

**Comment by cbm755** [comment]
tests failures where not your fault, I restarted them earlier: passed now.

I filed #10472 for the more general issue, because now I'm thinking we are correct.

I'm +1.  But I won't merge this until someone chimes in or #10472 is confirmed.

**Comment by hargup** [comment]
I'm +1 with having baseline in the middle.

**Comment by cbm755** [comment]
Ok, I plan to merge this in 24 hours---be bold right? :)

**Comment by asmeurer** [comment]
Yes, this printing (based on the added tests here) looks good. We need to fix the other printing so that things are centered properly.

**Comment by tokencolour** [comment]
Thank you, @cbm755 and @asmeurer !

**Pull request merged**